src/cluster_migrateslots.c

```diff
@@ -0,0 +1,1954 @@
+    listRelease(job->slot_ranges);
+    sdsfree(job->slot_ranges_str);
+    sdsfree(job->status_msg);
+    sdsfree(job->response_buf);
+    zfree(o);
+}
+
+void initClusterSlotMigrationJobList(void) {
+    server.cluster->slot_migration_jobs = listCreate();
+    listSetFreeMethod(server.cluster->slot_migration_jobs, freeSlotMigrationJob);
+}
+
+/* Convert a slotMigrationJobState enum to a user presentable string. */
+const char *slotMigrationJobStateToString(slotMigrationJobState state) {
+    switch (state) {
+    case SLOT_IMPORT_WAIT_ACK: return "waiting-for-ack";
+    case SLOT_IMPORT_RECEIVE_SNAPSHOT: return "receiving-snapshot";
+    case SLOT_IMPORT_WAITING_FOR_PAUSED: return "waiting-for-paused";
+    case SLOT_IMPORT_FAILOVER_REQUESTED: return "failover-requested";
+    case SLOT_IMPORT_FAILOVER_GRANTED: return "failover-granted";
+    case SLOT_IMPORT_FINISHED_WAITING_TO_CLEANUP: return "cleaning-up";
+
+    case SLOT_EXPORT_CONNECTING: return "connecting";
+    case SLOT_EXPORT_AUTHENTICATING: return "authenticating";
+    case SLOT_EXPORT_SEND_ESTABLISH: return "sending-establish-command";
+    case SLOT_EXPORT_READ_ESTABLISH_RESPONSE: return "reading-establish-response";
+    case SLOT_EXPORT_WAITING_TO_SNAPSHOT: return "waiting-to-snapshot";
+    case SLOT_EXPORT_SNAPSHOTTING: return "snapshotting";
+    case SLOT_EXPORT_STREAMING: return "replicating";
+    case SLOT_EXPORT_WAITING_TO_PAUSE: return "waiting-to-pause";
+    case SLOT_EXPORT_FAILOVER_PAUSED: return "failover-paused";
+    case SLOT_EXPORT_FAILOVER_GRANTED: return "failover-granted";
+
+    case SLOT_MIGRATION_JOB_SUCCESS: return "success";
+    case SLOT_MIGRATION_JOB_CANCELLED: return "cancelled";
+    case SLOT_MIGRATION_JOB_FAILED: return "failed";
+    }
+    return "unknown";
+}
+
+sds generateSlotMigrationJobDescription(slotMigrationJob *job, clusterNode *node) {
+    char *other_node_desc = job->type == SLOT_MIGRATION_EXPORT ? "target_node" : "source_node";
+    if (sdslen(node->human_nodename) > 0) {
+        return sdscatprintf(sdsempty(),
+                            "{name: %.40s, operation: %s, %s_id: %.40s, %s_human_name: %s, slots: %s}",
+                            job->name,
+                            job->type == SLOT_MIGRATION_EXPORT ? "export" : "import",
+                            other_node_desc, node->name, other_node_desc, node->human_nodename,
+                            job->slot_ranges_str);
+    } else {
+        return sdscatprintf(sdsempty(),
+                            "{name: %.40s, operation: %s, %s_id: %.40s, slots: %s}",
+                            job->name,
+                            job->type == SLOT_MIGRATION_EXPORT ? "export" : "import",
+                            other_node_desc, node->name,
+                            job->slot_ranges_str);
+    }
+}
+
+/* Update the provided job to the given state. */
+void updateSlotMigrationJobState(slotMigrationJob *job, slotMigrationJobState state) {
+    serverLog(LL_DEBUG,
```

Comment:
this log level seems low. can we bump it to NOTICE? the state transitions are going to be very useful in understanding operation issues imo.

```suggestion
    serverLog(LL_NOTICE,
```

Reply:
SGTM